### ISSUE TITLE
SonarQube and CodeQL analyze the base branch and not new code

### ISSUE THREAD

**Issue by I-Al-Istannen** [Together-Java/TJ-Bot#120]
**Describe the bug**
Currently the SonarQube and CodeQL workflows run with "pull_request_target" and will therefore see the state of the *base* of the branch and not the PR. Analyzing that code is not helpful.

**Expected behavior**
We want to analyze the *new code in the PR* instead

**Additional context**
Relevant reading material seems to be:
- Sonar:  
  Works apparently, but we currently build the code as well. This *must not happen*, otherwise any PR can steal the secrets that way. It looks like we can just omit the build though.
 https://community.sonarsource.com/t/github-comments-stopped-working-in-github-action-after-switch-to-pull-request-target/42556
~~- CodeQL seems *not to support* running on PRs from forks and should therefore probably be either removed or run on a schedule / push to develop. This is a bit late to find out any issues, but we can not run it on PRs from forks. See: https://github.com/github/codeql-action/pull/340#issuecomment-824691627~~

CodeQL does not need any secrets and can run in the `pull_request` scope.

**Comment by marko-radosavljevic** [comment]
Because: 
> TL;DR: Combining pull_request_target workflow trigger with an explicit checkout of an untrusted PR is a dangerous practice that may lead to repository compromise. - [link](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/)

Currently, it's not possible to make sonarcloud run on the PR code securely.

https://github.com/Together-Java/TJ-Bot/pull/121 makes sonar run only from a branch in the same repository, as a baindaid fix, until I find a way to also scan code from forks securely.

**Comment by I-Al-Istannen** [comment]
Until https://jira.sonarsource.com/browse/MMF-1371 is resolved we can probably stop trying to find a satisfying solution ourself.

**Comment by marko-radosavljevic** [comment]
Going a bit back to explain everything properly for documentation purposes.

Because SonarCloud workflow uses secrets, we do not want anyone from outside to be able to open malicious PR and extract our secrets.

To prevent that, we used `pull_request_target` trigger, that runs in the context of the target repository of the PR, rather than in the merge commit. This means the standard checkout action uses the target repository to prevent accidental usage of the user supplied code.

But, that means we do not have SonarCloud scans when we need them the most, inside PRs. So we explored some other ideas.
One of them is checking out PR explicitly, like this:

```yaml           
- name: Checkout
uses: actions/checkout@v2
with:
  ref: ${{ github.event.pull_request.head.sha }}
- name: SonarCloud Scan
uses: sonarsource/sonarcloud-github-action@v1.6
with:
  args:
  -Dsonar.pullrequest.key=${{ github.event.number }}
  -Dsonar.pullrequest.branch=${{ github.event.pull_request.head.ref }}
  -Dsonar.scm.revision=${{ github.event.pull_request.head.sha }}
```

But, that is also insecure:

> Warning: The pull_request_target event is granted a read/write repository token and can access secrets, even when it is triggered from a fork. Although the workflow runs in the context of the base of the pull request, you should make sure that you do not check out, build, or run untrusted code from the pull request with this event. Additionally, any caches share the same scope as the base branch, and to help prevent cache poisoning, you should not save the cache if there is a possibility that the cache contents were altered. For more information, see "Keeping your GitHub Actions and workflows secure: Preventing pwn requests" on the GitHub Security Lab website. - [link](https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#pull_request_target)


>TL;DR: Combining pull_request_target workflow trigger with an explicit checkout of an untrusted PR is a dangerous practice that may lead to repository compromise. - [link](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/)

So, currently, it's not possible to make SonarCloud run on the PR code securely. Sonar bug tracker: [MMF-1371](https://jira.sonarsource.com/browse/MMF-1371).

Because of that, https://github.com/Together-Java/TJ-Bot/pull/123 is merged, and we have linting inside PRs from forks with Sonarlint.

SonarCloud works on `pull_request` trigger, and it remains the main linting solution inside the project. It triggers after every merge, and triggers inside PRs that come from developers that have write permission. For PRs from forks, SonarCloud is not being run, and it fails. This is expected behavior.

That said, I'm closing this. We have #123 merged, that remedies this issue partially, and further progress is blocked by [MMF-1371](https://jira.sonarsource.com/browse/MMF-1371).
